### PR TITLE
Avkhokhlov/device tensor

### PR DESCRIPTION
**Pull request by AvKhokhlov**

### CODE REVIEW

File: Baikal/Kernels/CL/denoise.cl [hunk 1]
```diff
@@ -138,5 +138,46 @@ void DivideBySampleCount(GLOBAL float4* restrict dst,
...
+                     int dst_height,
+                     int dst_channels_offset, // offset inside pixel in channels (not bytes)
+                     int dst_channels_num,
+                     int src_width,
+                     int src_height,
+                     int src_channels_offset, // offset inside pixel in channels (not bytes)
+                     int src_channels_num,
+                     int channels_to_copy,
+                     int sample_count_enabled,
+                     GLOBAL float* out_sample_count)
```
Comment (dtarakanov1) [inline Baikal/Kernels/CL/denoise.cl]: `restrict`

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 2]
```diff
@@ -213,6 +223,43 @@ namespace Baikal
...
+            copy_kernel.SetArg(argc++, m_width);
+            copy_kernel.SetArg(argc++, m_height);
+            copy_kernel.SetArg(argc++, src_channels_offset);
+            copy_kernel.SetArg(argc++, src_channels_num);
+            copy_kernel.SetArg(argc++, channels_to_copy);
+            copy_kernel.SetArg(argc++, 1);
+            copy_kernel.SetArg(argc++, *m_sample_count_cache);
+
+            // run copy_kernel
+            {
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: Excessive. Same above.

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 3]
```diff
@@ -213,6 +223,43 @@ namespace Baikal
             }
         }
 
+
+        void MLDenoiser::WriteToTensor(CLWBuffer<RadeonRays::float3> src_buffer,
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: Let's call it `WriteToInputs`, and rename the `m_sample_count_cache` like `m_inputs_cache`.

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 4]
```diff
@@ -256,30 +302,17 @@ namespace Baikal
...
-                                                  *m_device_cache,
-                                                  m_host_cache.data(),
-                                                  device_mem.GetElementCount()).Wait();
-
-                    auto dest = host_mem;
-                    auto source = m_host_cache.data();
-                    sample_count = static_cast<unsigned int>(source->w);
+                    WriteToTensor(*m_device_cache, channels_count, 0, 4, 3);
+                    m_context->ReadBuffer<float>(0, *m_sample_count_cache, &real_sample_count, 1).Wait();
+                    sample_count = (unsigned)real_sample_count;
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: `static_cast<unsigned>()`

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 5]
```diff
@@ -213,6 +223,43 @@ namespace Baikal
...
+            copy_kernel.SetArg(argc++, m_height);
+            copy_kernel.SetArg(argc++, src_channels_offset);
+            copy_kernel.SetArg(argc++, src_channels_num);
+            copy_kernel.SetArg(argc++, channels_to_copy);
+            copy_kernel.SetArg(argc++, 1);
+            copy_kernel.SetArg(argc++, *m_sample_count_cache);
+
+            // run copy_kernel
+            {
+                auto thread_num = ((m_width * m_height + 63) / 64) * 64;
```
Comment (dtarakanov1) [inline Baikal/PostEffects/ML/denoiser.cpp]: Excessive brackets. Same above.

File: Baikal/Kernels/CL/denoise.cl [hunk 6]
```diff
@@ -138,5 +138,46 @@ void DivideBySampleCount(GLOBAL float4* restrict dst,
...
+                     int dst_width,
+                     int dst_height,
+                     int dst_channels_offset, // offset inside pixel in channels (not bytes)
+                     int dst_channels_num,
+                     int src_width,
+                     int src_height,
+                     int src_channels_offset, // offset inside pixel in channels (not bytes)
+                     int src_channels_num,
+                     int channels_to_copy,
+                     int sample_count_enabled,
```
Comment (achernigin1987) [inline Baikal/Kernels/CL/denoise.cl]: Lets do it always

File: Baikal/Kernels/CL/denoise.cl (hunk 1 above)
Comment (AvKhokhlov) [reply Baikal/Kernels/CL/denoise.cl]: Done

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 2 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 3 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 4 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 5 above)
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/Kernels/CL/denoise.cl (hunk 6 above)
Comment (AvKhokhlov) [reply Baikal/Kernels/CL/denoise.cl]: Done

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 7]
```diff
@@ -418,11 +408,13 @@ namespace Baikal
                 }
 
                 m_context->WriteBuffer<float3>(0,
+
                                                *m_last_denoised_image,
                                                m_host_cache.data(),
                                                inference_res.size() / 3);
                 m_has_denoised_image = true;
 
+
```
[the remark this answers is not in the document]
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done - the name is WriteToInputs

File: Baikal/PostEffects/ML/denoiser.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 8]
```diff
@@ -213,6 +223,43 @@ namespace Baikal
...
+            copy_kernel.SetArg(argc++, m_width);
+            copy_kernel.SetArg(argc++, m_height);
+            copy_kernel.SetArg(argc++, dst_channels_offset);
+            copy_kernel.SetArg(argc++, static_cast<int>(shape.channels));
+            copy_kernel.SetArg(argc++, m_width);
+            copy_kernel.SetArg(argc++, m_height);
+            copy_kernel.SetArg(argc++, src_channels_offset);
+            copy_kernel.SetArg(argc++, src_channels_num);
+            copy_kernel.SetArg(argc++, channels_to_copy);
+            copy_kernel.SetArg(argc++, 1);
```
[the remark this answers is not in the document]
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Deleted

File: Baikal/PostEffects/ML/denoiser.cpp [hunk 9]
```diff
@@ -213,6 +223,43 @@ namespace Baikal
...
+
+            int argc = 0;
+            copy_kernel.SetArg(argc++, *m_device_tensor);
+            copy_kernel.SetArg(argc++, src_buffer);
+            copy_kernel.SetArg(argc++, m_width);
+            copy_kernel.SetArg(argc++, m_height);
+            copy_kernel.SetArg(argc++, dst_channels_offset);
+            copy_kernel.SetArg(argc++, static_cast<int>(shape.channels));
+            copy_kernel.SetArg(argc++, m_width);
+            copy_kernel.SetArg(argc++, m_height);
```
[the remark this answers is not in the document]
Comment (AvKhokhlov) [reply Baikal/PostEffects/ML/denoiser.cpp]: Done

**Pull request merged**